invokeai/app/services/config.py

```diff
@@ -0,0 +1,364 @@
+            init_settings,
+            env_settings,
+            file_secret_settings,
+        ):
+            return (
+                init_settings,
+                cls._omegaconf_settings_source,
+                env_settings,
+                file_secret_settings,
+            )
+
+        @classmethod
+        def _omegaconf_settings_source(cls, settings: BaseSettings) -> dict[str, Any]:
+            if initconf := InvokeAISettings.initconf:
+                return initconf.get(settings.cmd_name(),{})
+            else:
+                return {}
+
+    @staticmethod
+    def add_field_argument(command_parser, name: str, field, default_override = None):
+        default = default_override if default_override is not None else field.default if field.default_factory is None else field.default_factory()
+        if get_origin(field.type_) == Literal:
+            allowed_values = get_args(field.type_)
+            allowed_types = set()
+            for val in allowed_values:
+                allowed_types.add(type(val))
+            allowed_types_list = list(allowed_types)
+            field_type = allowed_types_list[0] if len(allowed_types) == 1 else Union[allowed_types_list]  # type: ignore
+
+            command_parser.add_argument(
+                f"--{name}",
+                dest=name,
+                type=field_type,
+                default=default,
+                choices=allowed_values,
+                help=field.field_info.description,
+            )
+        else:
+            command_parser.add_argument(
+                f"--{name}",
+                dest=name,
+                type=field.type_,
+                default=default,
+                action=argparse.BooleanOptionalAction if field.type_==bool else 'store',
+                help=field.field_info.description,
+            )
+def _find_root()->Path:
+    if os.environ.get("INVOKEAI_ROOT"):
+        root = Path(os.environ.get("INVOKEAI_ROOT")).resolve()
+    elif (
+            os.environ.get("VIRTUAL_ENV")
+            and (Path(os.environ.get("VIRTUAL_ENV"), "..", INIT_FILE).exists()
+                 or
+                 Path(os.environ.get("VIRTUAL_ENV"), "..", LEGACY_INIT_FILE).exists()
+                 )
+    ):
+        root = Path(os.environ.get("VIRTUAL_ENV"), "..").resolve()
+    else:
+        root = Path("~/invokeai").expanduser().resolve()
+    return root
+
+class InvokeAIAppConfig(InvokeAISettings):
+    '''
+    Application-wide settings.
+    '''
+    #fmt: off
+    type: Literal["globals"] = "globals"
+    root                : Path = Field(default=_find_root(), description='InvokeAI runtime root directory')
+    infile              : Path = Field(default=None, description='Path to a file of prompt commands to bulk generate from')
+    precision           : Literal[tuple(['auto','float16','float32','autocast'])] = 'float16'
+    conf_path           : Path = Field(default='configs/models.yaml', description='Path to models definition file')
```

Comment:
I feel like some categorization might be useful for these settings. There's a lot of these that are model related, or file path related, and might make sense to group together.

Reply:
Done. If you add a `category` attribute to the `Field()`, the argparse help text will use it to group:
```
nvokeai-node-cli --help
usage: globals [-h] [--root ROOT] [--infile INFILE] [--conf_path CONF_PATH] [--model MODEL] [--outdir OUTDIR] [--embedding_dir EMBEDDING_DIR] [--lora_dir LORA_DIR]
               [--autoconvert_dir AUTOCONVERT_DIR] [--gfpgan_model_dir GFPGAN_MODEL_DIR] [--embeddings | --no-embeddings] [--xformers_enabled | --no-xformers_enabled]
               [--sequential_guidance | --no-sequential_guidance] [--precision {auto,float16,float32,autocast}] [--max_loaded_models MAX_LOADED_MODELS]
               [--always_use_cpu | --no-always_use_cpu] [--free_gpu_mem | --no-free_gpu_mem] [--nsfw_checker | --no-nsfw_checker] [--restore | --no-restore]
               [--esrgan | --no-esrgan] [--patchmatch | --no-patchmatch] [--internet_available | --no-internet_available] [--log_tokenization | --no-log_tokenization]

Application-wide settings.

options:
  -h, --help            show this help message and exit

Paths:
  --root ROOT           InvokeAI runtime root directory
  --infile INFILE       Path to a file of prompt commands to bulk generate from
  --conf_path CONF_PATH
                        Path to models definition file
  --outdir OUTDIR       Default folder for output images
  --embedding_dir EMBEDDING_DIR
                        Path to InvokeAI textual inversion aembeddings directory
  --lora_dir LORA_DIR   Path to InvokeAI LoRA model directory
  --autoconvert_dir AUTOCONVERT_DIR
                        Path to a directory of ckpt files to be converted into diffusers and imported on startup.
  --gfpgan_model_dir GFPGAN_MODEL_DIR
                        Path to GFPGAN models directory.

Models:
  --model MODEL         Initial model name
  --embeddings, --no-embeddings
                        Load contents of embeddings directory (default: True)

Memory/Performance:
  --xformers_enabled, --no-xformers_enabled
                        Enable/disable memory-efficient attention (default: True)
  --sequential_guidance, --no-sequential_guidance
                        Whether to calculate guidance in serial instead of in parallel, lowering memory requirements (default: False)
  --precision {auto,float16,float32,autocast}
                        Floating point precision
  --max_loaded_models MAX_LOADED_MODELS
                        Maximum number of models to keep in memory for rapid switching
  --always_use_cpu, --no-always_use_cpu
                        If true, use the CPU for rendering even if a GPU is available. (default: False)
  --free_gpu_mem, --no-free_gpu_mem
                        If true, purge model from GPU after each generation. (default: False)

Features:
  --nsfw_checker, --no-nsfw_checker
                        Enable/disable the NSFW checker (default: True)
  --restore, --no-restore
                        Enable/disable face restoration code (default: True)
  --esrgan, --no-esrgan
                        Enable/disable upscaling code (default: True)

```

Reply:
Will this also group it in config files? E.g. could I do:

```
global:
  models:
    root: "C:\my\sd\dir"
```